servicetalk-loadbalancer-experimental/src/main/java/io/servicetalk/loadbalancer/ConnectionPoolStrategy.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright © 2024 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.loadbalancer;
+
+import io.servicetalk.client.api.LoadBalancedConnection;
+
+import java.util.List;
+import java.util.function.Predicate;
+import javax.annotation.Nullable;
+
+/**
+ * A strategy for selecting connections at the {@link Host} level connection pool.
+ * @param <C> the concrete type of the connection.
+ */
+interface ConnectionPoolStrategy<C extends LoadBalancedConnection> {
```

Comment:
`ConnectionPoolStrategy` is a return type for public `ConnectionPoolStrategyFactory.buildStrategy`. Therefore, it should also be public.

Reply:
This was deliberate, but perhaps in bad taste. I don't want users to be able to define their own strategies, at least for now. Keeping this interface private means users can't really implement the `ConnectionPoolStrategyFactory` method, thus controlling proliferation.

Alternatively, we could make this an abstract class with package private constructor. Would that be preferable?

Reply:
Unfortunately, we have to make it public if it's a return type for another public method.

Interfaces usually much better than abstract classes. If you really want to hide it, then we should use a different strategy. For example, expose only a "config object" to users as public, but hide all implementation interfaces.